### PR TITLE
feat(cohort-query): New cohort query class

### PR DESCRIPTION
**Pull request by EDsCODE**
## Changes

- query that takes behavioral and person filters to create a cohort query

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: ee/clickhouse/queries/cohort_query.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    def get_sequence_column(self, sequence: Tuple[Event, Event, Period]) -> ColumnName:
+        return self.sequences_to_query_column_names[sequence]
+
+    def person_columns_to_query(self) -> Set[ColumnName]:
+        return set(
+            property_name for property_name, _, _, _ in self._used_properties_with_type_and_operator("person", "OR")
+        )
+
+    @cached_property
+    def is_using_person_properties_in_or(self) -> bool:
```
Comment (EDsCODE) [inline ee/clickhouse/queries/cohort_query.py:142]: When person-on-events is complete, we'll only need to join person table when a person filter is in an OR clause, otherwise we can use the person properties directly on events table to filter

**Comment by EDsCODE** [comment]
- [x] handle entity parameters when select fields are made
- [ ] write tests
- [x] make sure person prop push downs are handled (when all the filters are "AND") 
- [ ] ....

File: ee/clickhouse/queries/cohort_query.py (hunk 1 above)
Comment (rcmarron) [reply ee/clickhouse/queries/cohort_query.py:160]: I'd worry some about using the person on events to filter the persons. I think because of the time-based nature of the person-on-events info, we still want to use the actual person object when filtering persons.

Comment (neilkakkar) [reply ee/clickhouse/queries/cohort_query.py:160]: Rick, that's the breaking change we are okay with making when person-on-events comes, to get all those perf advances

File: ee/clickhouse/queries/cohort_query.py [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+NOW: Relative_Date = (0, "now")
+
+Period = Tuple[Relative_Date, Relative_Date]
+Event = Tuple[str, Tuple[str, int]]
+Event_In_Period = Tuple[Event, Period]
+
+USES_PERIOD = ["performed_event", "performed_event_muliple", "stopped_performing_event", "restarted_performing_event"]
+
+
+class CohortOptimizer(ColumnOptimizer):
```
Comment (neilkakkar) [inline ee/clickhouse/queries/cohort_query.py:27]: Q: Why is this related to `ColumnOptimizer`?

It seems like its mostly getting the column names to query, but also these column names are ones we're creating ourselves specifically for cohort filtering only

File: ee/clickhouse/queries/cohort_query.py [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+        _fields.extend(first_time_activity_fields)
+        _fields.extend(sequence_fields)
+
+        events = self._cohort_optimizer.get_events_to_query()
+
+        query = f"""
+        SELECT {", ".join(_fields)} FROM events {self.EVENT_TABLE_ALIAS}
+        {self._get_distinct_id_query()}
+        WHERE team_id = %(team_id)s
+        AND event IN %(events)s
```
Comment (neilkakkar) [inline ee/clickhouse/queries/cohort_query.py:251]: We always want a date filter in here! If each field has its own date range, this should be the max of all of them.

File: ee/clickhouse/queries/cohort_query.py [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+from posthog.models.filters.retention_filter import RetentionFilter
+from posthog.models.filters.session_recordings_filter import SessionRecordingsFilter
+from posthog.models.filters.stickiness_filter import StickinessFilter
+from posthog.models.property import Property, PropertyGroup, PropertyName
+from posthog.queries.column_optimizer import ColumnOptimizer
+
+Relative_Date = Tuple[int, str]
+NOW: Relative_Date = (0, "now")
+
+Period = Tuple[Relative_Date, Relative_Date]
```
Comment (neilkakkar) [inline ee/clickhouse/queries/cohort_query.py:20]: Q: What was the final decision on representing periods and dates?

File: ee/clickhouse/queries/cohort_query.py [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+
+    def _get_person_query(self) -> Tuple[str, Dict]:
+        """
+        FULL OUTER JOIN because the query needs to account for all people if there are or groups
+        """
+        if self._should_join_persons:
+            person_query, params = self._person_query.get_query()
+            return (
+                f"""
+            FULL OUTER JOIN ({person_query}) {self.PERSON_TABLE_ALIAS}
```
Comment (neilkakkar) [inline ee/clickhouse/queries/cohort_query.py:265]: This seems necessary only for a few specific combinations of operators?

Why not do something like we did for property pushdowns with and-or filters to determine what should happen here?

https://github.com/PostHog/posthog/blob/master/posthog/queries/property_optimizer.py#L1

Actually, we can reuse the above, if we change the query shape a bit.

File: ee/clickhouse/queries/cohort_query.py [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+                    and prop.event_type is not None
+                    and prop.time_interval is not None
+                    and prop.time_value is not None
+                ):
+                    _start_date = (prop.time_value, prop.time_interval)
+                    _end_date = NOW
+                    _period = (_start_date, _end_date)
+                    _event = (prop.event_type, prop.key)
+                    self.add_event_type(_event)
+                    _events[
```
Comment (neilkakkar) [inline ee/clickhouse/queries/cohort_query.py:64]: this data structure isn't making much sense to me right now, judging by how they're being used to get column name based on a period. I guess, question is, why do we have a column like this for each event/period combination?

Comment (neilkakkar) [reply ee/clickhouse/queries/cohort_query.py:64]: Actually, never mind the second part: That's clever, makes sense now. Instead of an array, it's columns, that can be turned into an array if necessary later on. Makes querying easier. Nice.

**Comment by neilkakkar** [comment]
Hmm, I like this direction, it's going to make things a lot faster than the naive one. In this case, I wonder if we should do this bit at all: https://github.com/PostHog/posthog/pull/9423/files#diff-9b66b50029d55fe5ff4b8fbf68ed490d125178d31660544741ee078602563d52R273-R288 ? 

Since it seems like nothing would use it right now, and adding this in here feels premature. Not sure what the general shape of behavioural filters would be like, when we introduce them everywhere, but this doesn't seem optimal for that either. (Maybe we'll end up using the cohort generator inline, which seems a lot more optimal / reasonable).

**Comment by EDsCODE** [comment]
> In this case, I wonder if we should do this bit at all: https://github.com/PostHog/posthog/pull/9423/files#diff-9b66b50029d55fe5ff4b8fbf68ed490d125178d31660544741ee078602563d52R273-R288 ?

Yup we should remove this

File: ee/clickhouse/queries/cohort_query.py (hunk 3 above)
Comment (EDsCODE) [reply ee/clickhouse/queries/cohort_query.py:251]: good point

File: ee/clickhouse/queries/cohort_query.py (hunk 1 above)
Comment (rcmarron) [reply ee/clickhouse/queries/cohort_query.py:160]: I agree with you when we're talking about filtering events on persons properties, but I think there's a nuance here when we're filtering persons. If you filter persons based on the person-on-events properties, then the user is going to be seeing inconsistent results based on the specific events that happen to be used in the query.

Comment (neilkakkar) [reply ee/clickhouse/queries/cohort_query.py:160]: Ah, I see what you mean!

The latter I think is what users might actually want: "Filter on people who had paid_status=unpaid at the time this event occurred".

Agree that there's some ambiguity here, and _maybe_ joining might not be the thing we want, in the regular OR case as well.. But we can cross this when we get to person-on-events.

File: ee/clickhouse/queries/cohort_query.py [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+            res, params = self.get_stopped_performing_event(prop, prepend, idx)
+        elif prop.value == "restarted_performing_event":
+            res, params = self.get_restarted_performing_event(prop, prepend, idx)
+        elif prop.value == "performed_event_first_time":
+            res, params = self.get_performed_event_first_time(prop, prepend, idx)
+        elif prop.value == "performed_event_sequence":
+            res, params = self.get_performed_event_sequence(prop, prepend, idx)
+        elif prop.value == "performed_event_regularly":
+            res, params = self.get_performed_event_regularly(prop, prepend, idx)
+        elif prop.value == "person":
```
Comment (neilkakkar) [inline ee/clickhouse/queries/cohort_query.py:173]: 🙈 

```suggestion
        elif prop.type == "person":
```

File: posthog/models/property.py [hunk 8]
```diff
@@ -164,7 +170,7 @@ def __init__(
             if getattr(self, key, None) is None:
                 raise ValueError(f"Missing required key {key} for property type {self.type}")
 
-        if self.type == "behavioural":
+        if self.type == "behavioural" and isinstance(self.value, BehaviouralPropertyType):
```
Comment (neilkakkar) [inline posthog/models/property.py:173]: shouldn't we error out when value is gibberish?

File: ee/clickhouse/models/property.py [hunk 9]
```diff
@@ -280,7 +280,7 @@ def prop_filter_json_extract(
     prop_var: str = "properties",
     allow_denormalized_props: bool = True,
     transform_expression: Optional[Callable[[str], str]] = None,
-    property_operator: PropertyOperatorType = PropertyOperatorType.AND,
+    property_operator: str = PropertyOperatorType.AND,
```
Comment (neilkakkar) [inline ee/clickhouse/models/property.py:283]: 🤔

File: ee/clickhouse/models/cohort.py [hunk 10]
```diff
@@ -163,7 +163,7 @@ def get_entity_cohort_subquery(cohort: Cohort, cohort_group: Dict, group_idx: in
         return f"distinct_id IN ({extract_person})", {**entity_params, **date_params}
 
 
-def _get_count_operator(count_operator: Optional[str]) -> str:
+def get_count_operator(count_operator: Optional[str]) -> str:
```
Comment (neilkakkar) [inline ee/clickhouse/models/cohort.py:166]: this is a bit annoying, as the property operator and cohort count operator are sadly disjoint

(property operator is `gt`, cohort operator is `gte`. Let's reconcile!

File: ee/clickhouse/queries/cohort_query.py [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+        self.params.update(condition_params)
+
+        behavior_subquery, behavior_subquery_params = self._get_behavior_subquery()
+        self.params.update(behavior_subquery_params)
+
+        person_query, person_params = self._get_person_query()
+        self.params.update(person_params)
+
+        final_query = f"""
+        SELECT person_id FROM
```
Comment (neilkakkar) [inline ee/clickhouse/queries/cohort_query.py:115]: Most upstream stuff appears to assume this is `id`

```suggestion
        SELECT person_id as id FROM
```

**Review by neilkakkar — APPROVED** [review]

File: ee/clickhouse/queries/cohort_query.py [hunk 12]
```diff
@@ -0,0 +1,573 @@
...
+        seq_date_param = f"{prepend}_seq_date_{idx}"
+        seq_date_interval = validate_interval(prop.seq_time_interval)
+
+        self._check_earliest_date([(date_value, date_interval), (seq_date_value, seq_date_interval)])
+
+        before_field = f"countIf(timestamp > now() - INTERVAL %({date_param})s {date_interval} - INTERVAL %({seq_date_param})s {seq_date_interval} AND timestamp < now() - INTERVAL %({date_param})s {date_interval} AND {entity_query}) AS {before_column_name}"
+        after_field = f"countIf(timestamp > now() - INTERVAL %({date_param})s {date_interval} AND timestamp < now() AND {entity_query}) AS {after_column_name}"
+
+        self._fields.append(before_field)
+        self._fields.append(after_field)
```
Comment (rcmarron) [inline ee/clickhouse/queries/cohort_query.py:380]: I think we need to also check that the event was initially performed prior to the time period of the `before field`.

File: ee/clickhouse/queries/cohort_query.py [hunk 13]
```diff
@@ -0,0 +1,573 @@
...
+
+        seq_date_value = parse_and_validate_positive_integer(prop.seq_time_value, "time_value")
+        seq_date_param = f"{prepend}_seq_date_{idx}"
+        seq_date_interval = validate_interval(prop.seq_time_interval)
+
+        self._check_earliest_date([(date_value, date_interval), (seq_date_value, seq_date_interval)])
+
+        before_field = f"countIf(timestamp > now() - INTERVAL %({date_param})s {date_interval} - INTERVAL %({seq_date_param})s {seq_date_interval} AND timestamp < now() - INTERVAL %({date_param})s {date_interval} AND {entity_query}) AS {before_column_name}"
+        after_field = f"countIf(timestamp > now() - INTERVAL %({date_param})s {date_interval} AND timestamp < now() AND {entity_query}) AS {after_column_name}"
+
```
Comment (rcmarron) [inline ee/clickhouse/queries/cohort_query.py:378]: The earliest date here should be an "all time" value.

File: ee/clickhouse/queries/cohort_query.py [hunk 14]
```diff
@@ -0,0 +1,573 @@
...
+            {date_condition}
+            GROUP BY person_id
+            """
+
+            query, params = (query, {"team_id": self._team_id, "events": self._events, **date_params,})
+
+        return query, params, self.BEHAVIOR_QUERY_ALIAS
+
+    def _get_persons_query(self) -> Tuple[str, Dict[str, Any], str]:
+        query, params = "", {}
```
Comment (rcmarron) [inline ee/clickhouse/queries/cohort_query.py:237]: For `restarted_performing_event` and `performed_event_first_time`, the earliest date is "all time", so we should support that concept here. Makes me wonder if the earliest date should be event specific. Might be overoptimizing, so just a thought.

**Review by rcmarron — COMMENTED** [review]
Just added a few failing tests. Pretty sure the comments highlight the cause of the failures.

File: ee/clickhouse/queries/cohort_query.py [hunk 15]
```diff
@@ -0,0 +1,573 @@
...
+
+        seq_date_value = parse_and_validate_positive_integer(prop.seq_time_value, "time_value")
+        seq_date_param = f"{prepend}_seq_date_{idx}"
+        seq_date_interval = validate_interval(prop.seq_time_interval)
+
+        self._check_earliest_date([(date_value, date_interval), (seq_date_value, seq_date_interval)])
+
+        before_field = f"countIf(timestamp > now() - INTERVAL %({date_param})s {date_interval} - INTERVAL %({seq_date_param})s {seq_date_interval} AND timestamp < now() - INTERVAL %({date_param})s {date_interval} AND {entity_query}) AS {before_column_name}"
+        after_field = f"countIf(timestamp > now() - INTERVAL %({date_param})s {date_interval} AND timestamp < now() AND {entity_query}) AS {after_column_name}"
+
```
Comment (rcmarron) [inline ee/clickhouse/queries/cohort_query.py:405]: The earliest date here should be an "all time" value, and we should check that the event occurred once before the `before field`. Otherwise, it isn't necessarily a "restarting" condition.

File: ee/clickhouse/queries/cohort_query.py [hunk 16]
```diff
@@ -0,0 +1,573 @@
...
+                date_query += f" - INTERVAL %({param})s {validate_interval(date[1])}"
+                date_params = {**date_params, param: date[0]}
+
+        return date_query, date_params
+
+    def _check_earliest_date(self, relative_dates: List[Relative_Date]) -> None:
+        if not self._earliest_time:
+            self._earliest_time = relative_dates
+        elif relative_date_is_greater(relative_dates, self._earliest_time):
+            self._earliest_time = relative_dates
```
Comment (rcmarron) [inline ee/clickhouse/queries/cohort_query.py:264]: For `restarted_performing_event` and `performed_event_first_time`, the earliest date is "all time", so we should support that concept here. Makes me wonder if the earliest date should be event specific. Might be over optimizing, so just a thought.

**Comment by neilkakkar** [comment]
@rcmarron did you pushhh?

**Comment by rcmarron** [comment]
> @rcmarron did you pushhh?

gahhhh no! just did

**Comment by rcmarron** [comment]
💥
